pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if !t.repo.pullthrough {
+			dgst, err := digest.ParseDigest(history.Items[0].Image)
+			if err != nil {
+				return nil, err
+			}
+
+			image, err := t.repo.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			return tags, err
+		}
+
+		if !t.repo.pullthrough {
+			image, err := t.repo.getImage(dgst)
+			if err != nil {
+				return nil, err
+			}
+
+			if !isImageManaged(image) {
+				continue
+			}
+		}
+
+		if dgst == desc.Digest {
+			tags = append(tags, tag)
+		}
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	image, err := t.repo.registryOSClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("(tagService).Tag: Unable to get image: %s", dgst.Digest.String())
+		return err
+	}
+	image.SetResourceVersion("")
+
+	if t.repo.pullthrough && !isImageManaged(image) {
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+		Tag:   tag,
+		Image: *image,
+	}
+
+	return t.repo.registryOSClient.ImageStreamMappings(imageStream.Namespace).Create(&ism)
+}
+
+func (t tagService) Untag(ctx context.Context, tag string) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	te := imageapi.LatestTaggedImage(imageStream, tag)
+	if te == nil {
+		return distribution.ErrTagUnknown{Tag: tag}
+	}
+
+	if !t.repo.pullthrough {
+		dgst, err := digest.ParseDigest(te.Image)
+		if err != nil {
+			return err
+		}
+
+		image, err := t.repo.getImage(dgst)
+		if err != nil {
+			return err
```

Comment:
Handle `NotFound` in a way appropriate for distribution (`ManifestUnknown` or something like that).